crates/bevy_render/src/batching/mod.rs

```diff
@@ -115,6 +131,112 @@ pub fn batch_and_prepare_render_phase<I: CachedRenderPipelinePhaseItem, F: GetBa
     }
 }
 
+/// Sorts a render phase that uses bins.
+pub fn sort_binned_render_phase<BPI>(mut views: Query<&mut BinnedRenderPhase<BPI>>)
+where
+    BPI: BinnedPhaseItem,
+{
+    for mut phase in &mut views {
+        phase.batchable_keys.sort_unstable();
+        phase.unbatchable_keys.sort_unstable();
+    }
+}
+
+/// Creates batches for a render phase that uses bins.
+pub fn batch_and_prepare_binned_render_phase<BPI, GBBD>(
+    gpu_array_buffer: ResMut<GpuArrayBuffer<GBBD::BufferData>>,
+    mut views: Query<&mut BinnedRenderPhase<BPI>>,
+    param: StaticSystemParam<GBBD::Param>,
+) where
+    BPI: BinnedPhaseItem,
+    GBBD: GetBinnedBatchData,
+{
+    let gpu_array_buffer = gpu_array_buffer.into_inner();
+    let system_param_item = param.into_inner();
+
+    for mut phase in &mut views {
+        let phase = &mut *phase; // Borrow checker.
+
+        // Prepare batchables.
+
+        let mut is_first_instance = true;
+        for key in &phase.batchable_keys {
+            let mut batch: Option<BinnedRenderPhaseBatch> = None;
+            for &entity in &phase.batchable_values[key] {
+                let Some(buffer_data) = GBBD::get_batch_data(&system_param_item, entity) else {
+                    let instance_index = match phase.batches.last() {
+                        Some(batch) => batch.instance_end_index,
+                        None => 0,
+                    };
+                    phase
+                        .batches
+                        .push(BinnedRenderPhaseBatch::placeholder(instance_index));
```

Comment:
why it needed ? IMO,it's to maintain equality between the lengths of  `phase.batches` and  `phase.batchable_keys`, but it still causes inequality when two or more entities enter this branch.

Reply:
I fixed this bug.